astropy/time/core.py

```diff
@@ -874,12 +874,43 @@ def shape(self, shape):
                     reshaped.append(val)
 
     def _shaped_like_input(self, value):
-        out = value
-        if value.dtype.kind == 'M':
-            return value[()]
-        if not self._time.jd1.shape and not np.ma.is_masked(value):
-            out = value[()] if value.dtype.fields else value.item()
-        return out
+        if not isinstance(self._time.jd1, np.ndarray):
+            raise TypeError(
+                f"Time objects should always contain arrays internally but "
+                f"this one has {self._time.jd1} of type "
+                f"{type(self._time.jd1)} and {self._time.jd2} of type "
+                f"{type(self._time.jd2)}.")
+        elif self._time.jd1.shape:
+            if isinstance(value, np.ndarray):
+                out = value
+            else:
+                raise TypeError(
+                    f"JD is an array ({self._time.jd1!r}) but value "
+                    f"is not ({value!r})")
+        else:
+            # zero-dimensional array
+            if (isinstance(value, np.ndarray)
+                    and not value.shape
+                    and not np.ma.is_masked(value)):
+                out = value[()]
+                #out = value[()] if value.dtype.fields else value.item()
+            else:
+                out = value
+        # De-numpify types for the convenience of callers
+        if isinstance(out, np.float64):
+            return float(out)
+        elif isinstance(out, np.str_):
+            return str(out)
+        elif callable(out) and not hasattr(self._time.__class__.value, "fget"):
```

Comment:
The documentation is quite clear that `value` needs to be a property, so I think this test is not appropriate here and ask that you remove it.  A better place to do this validation is when a new format is added to the registry.  Maybe you can use your experience to add other tests at that point?

Reply:
It's a cleaner check there, thank you.